<a href="https://colab.research.google.com/github/james-caldwell1981/DS-Unit-1-Sprint-2-Statistics/blob/master/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
import pandas as pd
import numpy as np
from scipy import stats

In [0]:
#Create column header list

headers = ['symboling', 
           'aspiration', 
           'num-of-doors', 
           'body-style', 
           'drive-wheels', 
           'engine-location',
           'wheel-base',
           'length',
           'width',
           'height',
           'curb-weight',
           'engine-type',
           'fuel-system',
           'bore',
           'stroke',
           'compression-ratio',
           'horsepower',
           'peak-rpm',
           'city-mpg',
           'highway-mpg',
           'price']

In [0]:
#read in data set

df = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data', names=headers)

In [4]:
df.head()

symboling   aspiration  ... highway-mpg  price
3 ?   alfa-romero gas std       two  convertible  ...          27  13495
                      std       two  convertible  ...          27  16500
1 ?   alfa-romero gas std       two    hatchback  ...          26  16500
2 164 audi        gas std      four        sedan  ...          30  13950
                      std      four        sedan  ...          22  17450

[5 rows x 21 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [5]:
#sample data set for 10 vehicles

sample = df.sample(10, random_state=30)
sample.head()

symboling   aspiration  ... highway-mpg  price
 1 154 plymouth      gas std      four        sedan  ...          38   6692
-1 65  toyota        gas std      four        sedan  ...          32  10898
 1 148 dodge         gas std      four    hatchback  ...          38   6229
-1 137 mitsubishi    gas std      four        sedan  ...          30   9279
 3 142 mercedes-benz gas std       two  convertible  ...          18  35056

[5 rows x 21 columns]

Null Hypothesis: Sample highway fuel economy is similar to population highway fuel economy

Alternative Hypothesis: Sample highway fuel economy is different than population fuel economy.

Confidence Level: 95%

In [6]:
#get mean of all highway fuel economy

ttl_hmpg = df['highway-mpg'].mean()
ttl_hmpg

30.75121951219512

In [7]:
#get mean of sample highway fuel economy

smpl_hmpg = sample['highway-mpg'].mean()
smpl_hmpg

32.8

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [8]:
stats.ttest_1samp(df['highway-mpg'], .05, nan_policy='omit')

Ttest_1sampResult(statistic=63.83187358723746, pvalue=8.862423699126691e-137)

**Conclusion:**
Based on a t-statistic of 63.83 and a p-value of 8.86, I reject the null hypothesis that the sample set highway fuel economy properly represents highway fuel economy of the entire fleet of vehicles.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [9]:
sample2 = df.sample(100, random_state=30)
stats.ttest_1samp(sample2['highway-mpg'], 0.05)

Ttest_1sampResult(statistic=45.734324932718316, pvalue=2.1809947925859347e-68)

100 Car Conclusion:

Based on a t statistic of 45.73 and a p-value of 2.18, I still reject the null hypothesis of sample set highway fuel economy matches population highway fuel economy. The larger sample size indicates fuel economy in population is more efficient than what can be inferred from the original 10 vehicle sample.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The exponentially larger sample size gets a more accurate set of data points from sampling a higher percentage of the population.

A random sample has a small chance to yield results which are skewed in one direction or another if each subject has similar attributes.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

data_url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data'

df2 = pd.read_csv(data_url, names=column_headers)

print(df2.shape)
df2.head()

In [0]:
df2 = df2.replace({'y':1, 'n':0, '?':np.nan})
df2.head()

In [0]:
rep = df2[df2['party']=='republican']
dem = df2[df2['party']=='democrat']

In [18]:
rep.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

In [19]:
dem.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [0]:
samp1_dem = dem['el-salvador-aid']
samp1_rep = rep['el-salvador-aid']

In [0]:
samp2_dem = dem['budget']
samp2_rep = rep['budget']

In [105]:
stats.ttest_1samp(samp2_dem, .01, nan_policy='omit')

Ttest_1sampResult(statistic=44.90978673539431, pvalue=3.1056211436272872e-124)

In [106]:
stats.ttest_1samp(samp2_rep, .01, nan_policy='omit')

Ttest_1sampResult(statistic=4.650675773180401, pvalue=6.803300744518018e-06)

In [107]:
stats.ttest_ind(samp2_dem, samp2_rep, nan_policy='omit')

Ttest_indResult(statistic=23.21277691701378, pvalue=2.0703402795404463e-77)

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

Null Hypothesis: Republicans support El Salvador aid more than Democrats.

Alternative Hypothesis: Republicans do not support to El Salvador aid more than Democrats.

In [102]:
stats.ttest_ind(samp1_dem, samp1_rep, nan_policy='omit')

Ttest_indResult(statistic=-21.13669261173219, pvalue=5.600520111729011e-68)

In [103]:
samp1_rep.mean()

0.9515151515151515

Conclusion:

I fail to reject the Null Hypothesis of Republicans supporting El Salvador aid more than Democrats.

Null Hypothesis: Democrats support the budget more than Republicans.

Alternative Hypothesis: Democrats do not support the budget more than Republicans.

In [109]:
stats.ttest_ind(samp2_rep, samp2_dem, nan_policy='omit')

Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)

In [110]:
samp2_dem.mean()

0.8884615384615384

Conclustion:

I fail to reject the Null Hypothesis of Democrats supporting the budget more than Republicans.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [130]:
df2.dtypes

party                    object
handicapped-infants     float64
water-project           float64
budget                  float64
physician-fee-freeze    float64
el-salvador-aid         float64
religious-groups        float64
anti-satellite-ban      float64
aid-to-contras          float64
mx-missile              float64
immigration             float64
synfuels                float64
education               float64
right-to-sue            float64
crime                   float64
duty-free               float64
south-africa            float64
dtype: object

In [0]:
df3 = df2.drop('party', axis=1)

In [0]:
def test(df):
    tests = []
    for i in df3.columns:
        j=0
        for j in df3.columns:
            if j!=i:
                tp = stats.ttest_ind(df3[i], df3[j], nan_policy='omit')
                tests.append([i, j , tp[0], tp[1]])                
    return tests
                

results = test(df3)

In [157]:
results = pd.DataFrame(results)

results.head()

0                     1         2             3
0  handicapped-infants         water-project -1.761031  7.861130e-02
1  handicapped-infants                budget -4.552811  6.070477e-06
2  handicapped-infants  physician-fee-freeze  0.723291  4.697016e-01
3  handicapped-infants       el-salvador-aid -1.823945  6.851514e-02
4  handicapped-infants      religious-groups -5.937781  4.216407e-09

In [0]:
r_headers = ['topic_1', 'topic_2', 't_statistic', 'p_value']



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE